File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/motion/SpeedMotionManager.kt
package com.ikolvi.tracelet.sdk.motion
import com.ikolvi.tracelet.sdk.util.TraceletLog

import android.os.SystemClock
import android.util.Log
import com.ikolvi.tracelet.sdk.ConfigManager
import com.ikolvi.tracelet.sdk.StateManager
import com.ikolvi.tracelet.sdk.TraceletEventSender
import com.ikolvi.tracelet.sdk.model.SpeedMotionState
import com.ikolvi.tracelet.sdk.model.StationaryTrackingMode

/**
 * Speed-based motion detection state machine.
 *
 * Uses GPS speed from location fixes to transition between MOVING, SLOWING,
 * and STATIONARY states, switching the native location engine between
 * continuous tracking and low-power periodic/geofence fixes.
 *
 * This is a peer to [MotionDetector] (accelerometer-based) — the two are
 * mutually exclusive, selected via `MotionConfig.motionDetectionMode`.
 *
 * State machine:
 * ```
 * MOVING  --speed<threshold-->  SLOWING  --delay elapsed-->  STATIONARY
 *   ^                              |                             |
 *   |          speed>=threshold----+     wakeConfirmCount fixes--+
 *   +------------------------------------------------------------+
 * ```
 */
class SpeedMotionManager(
    private val config: ConfigManager,
    private val state: StateManager,
    var events: TraceletEventSender,
    private val callback: SpeedMotionCallback,
) {
    companion object {
        private const val TAG = "SpeedMotion"
    }
    /**
     * Callback interface for mode switching, implemented by the host
     * (LocationService or equivalent).
     */
    interface SpeedMotionCallback {
        fun switchToContinuous()
        fun switchToStationaryPeriodic()
        fun switchToStationaryGeofences()
    }

    // Current state
    private var currentState: SpeedMotionState = SpeedMotionState.MOVING
    private var lowSpeedCount: Int = 0
    private var wakeCount: Int = 0

    // Timing for SLOWING -> STATIONARY countdown
    private var lastFixTime: Long = 0L
    private var avgIntervalMs: Long = 0L
    private var fixCount: Long = 0L

    // Config values (cached on start)
    private var speedMovingThreshold: Double = 1.5
    private var speedStationaryDelay: Int = 180
    private var stationaryTrackingMode: StationaryTrackingMode = StationaryTrackingMode.PERIODIC
    private var speedWakeConfirmCount: Int = 1

    private var started = false

    // =========================================================================
    // Public API
    // =========================================================================

    /**
     * Start speed-based motion detection.
     *
     * Loads persisted state from [StateManager] so the correct mode is
     * resumed after process death / reboot.
     */
    fun start(forceMoving: Boolean = false) {
        if (started) return
        started = true

        // Cache config with bounds enforcement
        speedMovingThreshold = config.getSpeedMovingThreshold()
        stationaryTrackingMode = config.getStationaryTrackingMode()

        val rawDelay = config.getSpeedStationaryDelay()
        speedStationaryDelay = rawDelay.coerceAtLeast(0)
        if (rawDelay < 0) {
            TraceletLog.warning("speedStationaryDelay was $rawDelay, clamped to 0")
        } else if (rawDelay == 0) {
            TraceletLog.warning("speedStationaryDelay is 0 — device will transition to STATIONARY immediately after a single low-speed fix")
        }

        val rawWakeCount = config.getSpeedWakeConfirmCount()
        speedWakeConfirmCount = rawWakeCount.coerceAtLeast(1)
        if (rawWakeCount < 1) {
            TraceletLog.warning("speedWakeConfirmCount was $rawWakeCount, clamped to 1")
        }

        // Restore persisted state
        currentState = state.speedMotionState ?: SpeedMotionState.MOVING
        lowSpeedCount = state.speedLowCount
        wakeCount = state.speedWakeCount
        lastFixTime = state.speedLastTransition

        // If explicitly forced, ensure we start in MOVING state.
        if (forceMoving) {
            currentState = SpeedMotionState.MOVING
            lowSpeedCount = 0
            wakeCount = 0
            state.speedMotionState = currentState
            state.speedLowCount = 0
            state.speedWakeCount = 0
            state.isMoving = true
            TraceletLog.debug("start() — forced to MOVING state")
        } else {
            TraceletLog.debug("start() — restored state=$currentState, lowCount=$lowSpeedCount, wakeCount=$wakeCount")

            if (currentState == SpeedMotionState.STATIONARY) {
                switchToStationary()
            } else if (currentState == SpeedMotionState.SLOWING) {
                slowingStartTimeMs = android.os.SystemClock.elapsedRealtime()
                startSlowingTimer()
            }
        }
    }

    /** Stop speed-based motion detection. */
    fun stop() {
        if (!started) return
        started = false
        TraceletLog.debug("stop()")
    }

    /**
     * Handle manual pace changes triggered by the caller.
     */
    fun onManualPaceChange(isMoving: Boolean) {
        if (!started) return
        TraceletLog.debug("onManualPaceChange(isMoving=$isMoving)")
        if (isMoving) {
            lowSpeedCount = 0
            wakeCount = 0
            stopSlowingTimer()
            transitionTo(SpeedMotionState.MOVING)
            callback.switchToContinuous()
        } else {
            lowSpeedCount = 0
            wakeCount = 0
            stopSlowingTimer()
            transitionTo(SpeedMotionState.STATIONARY)
            switchToStationary()
        }
    }

    /** Returns the current state value string (legacy compatibility). */
    fun getCurrentState(): String = currentState.name.lowercase()

    /**
     * Feed a new location fix's speed into the state machine.
     *
     * Called by [LocationEngine] on every continuous or periodic fix
     * when speed-based motion detection is active.
     *
     * @param speedMetersPerSecond GPS speed from the location fix.
     */
    fun onLocation(speedMetersPerSecond: Double) {
        if (!started) return
        val speed = if (speedMetersPerSecond < 0) 0.0 else speedMetersPerSecond

        // Track inter-fix interval for SLOWING countdown.
        // Use elapsedRealtime() which is monotonic and immune to NTP / manual clock changes.
        val now = SystemClock.elapsedRealtime()
        if (lastFixTime > 0) {
            val interval = now - lastFixTime
            fixCount++
            avgIntervalMs = if (avgIntervalMs == 0L) {
                interval
            } else {
                // Exponential moving average
                (avgIntervalMs * 3 + interval) / 4
            }
        }
        lastFixTime = now

        when (currentState) {
            SpeedMotionState.MOVING -> onLocationMoving(speed)
            SpeedMotionState.SLOWING -> onLocationSlowing(speed)
            SpeedMotionState.STATIONARY -> onLocationStationary(speed)
        }
    }

    // =========================================================================
    // State handlers
    // =========================================================================

    private fun onLocationMoving(speed: Double) {
        if (speed < speedMovingThreshold) {
            TraceletLog.debug("MOVING -> SLOWING (speed=${formatSpeed(speed)} < threshold=$speedMovingThreshold)")
            lowSpeedCount = 1
            slowingStartTimeMs = SystemClock.elapsedRealtime()
            transitionTo(SpeedMotionState.SLOWING)
            startSlowingTimer()
        }
    }

    private var slowingStartTimeMs: Long = 0L
    private var slowingTimerRunnable: java.lang.Runnable? = null
    private val slowingHandler = android.os.Handler(android.os.Looper.getMainLooper())

    private fun startSlowingTimer() {
        stopSlowingTimer()
        val delayMs = speedStationaryDelay * 1000L
        slowingTimerRunnable = java.lang.Runnable {
            if (currentState == SpeedMotionState.SLOWING) {
                TraceletLog.debug("SLOWING timer expired -> STATIONARY")
                lowSpeedCount = 0
                wakeCount = 0
                transitionTo(SpeedMotionState.STATIONARY)
                switchToStationary()
            }
        }
        slowingHandler.postDelayed(slowingTimerRunnable!!, delayMs)
    }

    private fun stopSlowingTimer() {
        slowingTimerRunnable?.let { slowingHandler.removeCallbacks(it) }
        slowingTimerRunnable = null
    }

    private fun switchToStationary() {
        when (stationaryTrackingMode) {
            StationaryTrackingMode.GEOFENCES -> {
                TraceletLog.debug("Switching to stationary geofences mode")
                callback.switchToStationaryGeofences()
            }
            else -> {
                TraceletLog.debug("Switching to stationary periodic mode")
                callback.switchToStationaryPeriodic()
            }
        }
    }

    private fun onLocationSlowing(speed: Double) {
        if (speed >= speedMovingThreshold) {
            TraceletLog.debug("SLOWING -> MOVING (speed=${formatSpeed(speed)} >= threshold=$speedMovingThreshold)")
            lowSpeedCount = 0
            stopSlowingTimer()
            transitionTo(SpeedMotionState.MOVING)
            return
        }

        lowSpeedCount++
        state.speedLowCount = lowSpeedCount

        val elapsedMs = SystemClock.elapsedRealtime() - slowingStartTimeMs
        val delayMs = speedStationaryDelay * 1000L

        TraceletLog.debug("SLOWING: lowCount=$lowSpeedCount, elapsed=${elapsedMs}ms, delay=${delayMs}ms, speed=${formatSpeed(speed)}")

        if (elapsedMs >= delayMs) {
            TraceletLog.debug("SLOWING -> STATIONARY (elapsed ${elapsedMs}ms >= delay ${delayMs}ms)")
            lowSpeedCount = 0
            wakeCount = 0
            stopSlowingTimer()
            transitionTo(SpeedMotionState.STATIONARY)
            switchToStationary()
        }
    }

    private fun onLocationStationary(speed: Double) {
        if (speed >= speedMovingThreshold) {
            wakeCount++
            state.speedWakeCount = wakeCount
            TraceletLog.debug("STATIONARY: wake fix (speed=${formatSpeed(speed)}), wakeCount=$wakeCount/$speedWakeConfirmCount")

            if (wakeCount >= speedWakeConfirmCount) {
                TraceletLog.debug("STATIONARY -> MOVING (wakeCount=$wakeCount >= confirm=$speedWakeConfirmCount)")
                wakeCount = 0
                transitionTo(SpeedMotionState.MOVING)
                callback.switchToContinuous()
            }
        } else {
            // Low speed — reset wake counter, stay stationary
            if (wakeCount > 0) {
                TraceletLog.debug("STATIONARY: low speed, resetting wakeCount ($wakeCount -> 0)")
            }
            wakeCount = 0
            state.speedWakeCount = 0
        }
    }

    // =========================================================================
    // State transition + persistence + event emission
    // =========================================================================

    private fun transitionTo(newState: SpeedMotionState) {
        val previousState = currentState
        currentState = newState

        // Persist to SharedPreferences
        state.speedMotionState = newState
        state.speedLowCount = lowSpeedCount
        state.speedWakeCount = wakeCount
        state.speedLastTransition = SystemClock.elapsedRealtime()

        // Update isMoving for compatibility with existing consumers
        state.isMoving = newState != SpeedMotionState.STATIONARY

        // Emit speed motion change event
        val eventData = mapOf<String, Any>(
            "state" to newState.ordinal,
            "previousState" to previousState.ordinal,
            "trackingMode" to when (newState) {
                SpeedMotionState.STATIONARY -> if (stationaryTrackingMode == StationaryTrackingMode.GEOFENCES) 1 else 2
                else -> 0
            }
        )
        events.sendSpeedMotionChange(eventData)

        TraceletLog.debug("State transition: ${previousState.name} -> ${newState.name}")
    }

    private fun formatSpeed(speed: Double): String = "%.2f m/s".format(speed)
}

File: sdk/ios/Sources/TraceletSDK/motion/SpeedMotionManager.swift
import Foundation

/// Delegate that the `SpeedMotionManager` calls to switch the location engine
/// between continuous and stationary modes and to emit Pigeon events.
public protocol SpeedMotionDelegate: AnyObject {
    func switchToContinuous()
    func switchToStationaryPeriodic()
    func switchToStationaryGeofences()
    /// Emit a speed-motion state change event to Dart.
    func emitSpeedMotionEvent(state: Int, previousState: Int, trackingMode: Int)
}

/// GPS-speed-based motion detection state machine.
///
/// Drives MOVING -> SLOWING -> STATIONARY transitions based on
/// `CLLocation.speed` from each location fix. Designed for vehicle tracking
/// where accelerometer-based stop detection is unreliable (phone on a smooth
/// dashboard reads near-zero even at highway speed).
///
/// This class is a peer to `MotionDetector` (not a branch inside it).
/// Inputs are location fixes, not sensor streams.
///
/// ## State Machine
///
/// ```
/// MOVING (continuous) --speed<threshold--> SLOWING (continuous)
///    ^                                          |
///    |                                   delay elapsed
///    |                                          v
///    +--wakeConfirmCount fixes>=threshold-- STATIONARY (periodic or geofences)
/// ```
public final class SpeedMotionManager {

    // MARK: - State

    public enum SpeedMotionState: Int {
        case moving = 0
        case slowing = 1
        case stationary = 2
    }

    public private(set) var state: SpeedMotionState = .moving

    /// Consecutive low-speed samples while SLOWING.
    public private(set) var lowSpeedCount: Int = 0

    /// Consecutive wake-speed samples while STATIONARY.
    public private(set) var wakeCount: Int = 0

    /// Timestamp of the first low-speed sample that initiated the current
    /// SLOWING phase. Used for elapsed-time calculation.
    private var slowingStartTime: TimeInterval = 0

    /// Timestamp of the last `onLocation` call, for approximate inter-fix interval.
    private var lastFixTime: TimeInterval = 0

    private var isRunning = false

    // MARK: - Config

    /// Speed (m/s) below which the device is considered not moving.
    public var speedMovingThreshold: Double = 1.5

    /// Seconds of sustained low speed before transitioning to STATIONARY.
    public var speedStationaryDelay: Int = 180

    /// Stationary tracking mode: periodic or geofences.
    public var stationaryTrackingMode: StationaryTrackingMode = .periodic

    /// Interval for periodic one-shot fixes in stationary mode (seconds).
    public var stationaryPeriodicInterval: Int = 120

    /// Number of consecutive high-speed fixes required to wake from STATIONARY.
    public var speedWakeConfirmCount: Int = 1

    // MARK: - Dependencies

    public weak var delegate: SpeedMotionDelegate?

    private let stateManager: StateManager

    // MARK: - Init

    public init(stateManager: StateManager) {
        self.stateManager = stateManager
    }

    // MARK: - Lifecycle

    /// Start the speed motion manager. Loads persisted state from StateManager.
    public func start(forceMoving: Bool = false) {
        guard !isRunning else { return }
        isRunning = true

        // Validate config bounds
        if speedStationaryDelay < 0 {
            NSLog("[SpeedMotion] WARNING: speedStationaryDelay was %d, clamping to 0", speedStationaryDelay)
            speedStationaryDelay = 0
        } else if speedStationaryDelay == 0 {
            NSLog("[SpeedMotion] WARNING: speedStationaryDelay is 0 — device will transition to STATIONARY immediately after a single low-speed fix")
        }
        if speedWakeConfirmCount < 1 {
            NSLog("[SpeedMotion] WARNING: speedWakeConfirmCount was %d, clamping to 1", speedWakeConfirmCount)
            speedWakeConfirmCount = 1
        }

        // Restore persisted state
        if let persisted = stateManager.speedMotionState,
           let restored = SpeedMotionState(rawValue: persisted) {
            state = restored
        } else {
            state = .moving
        }
        lowSpeedCount = stateManager.speedLowCount
        wakeCount = stateManager.speedWakeCount
        slowingStartTime = stateManager.speedLastTransition

        if forceMoving {
            state = .moving
            lowSpeedCount = 0
            wakeCount = 0
            stateManager.speedMotionState = state.rawValue
            stateManager.speedLowCount = 0
            stateManager.speedWakeCount = 0
            stateManager.isMoving = true
            NSLog("[SpeedMotion] start() — forced to MOVING state")
        } else {
            NSLog("[SpeedMotion] start: restored state=%d, lowSpeedCount=%d, wakeCount=%d",
                  state.rawValue, lowSpeedCount, wakeCount)
            
            if state == .stationary {
                switchToStationary()
            }
        }
    }

    /// Stop the speed motion manager and reset runtime counters.
    public func stop() {
        guard isRunning else { return }
        isRunning = false
        NSLog("[SpeedMotion] stop")
    }

    /// Handle manual pace changes triggered by the caller.
    public func onManualPaceChange(isMoving: Bool) {
        guard isRunning else { return }
        NSLog("[SpeedMotion] onManualPaceChange(isMoving=\(isMoving))")
        if isMoving {
            lowSpeedCount = 0
            wakeCount = 0
            stopSlowingTimer()
            let previousState = state
            state = .moving
            
            if state != previousState {
                persistState()
                emitEvent(previous: previousState, current: state)
            }
            delegate?.switchToContinuous()
        } else {
            lowSpeedCount = 0
            wakeCount = 0
            stopSlowingTimer()
            let previousState = state
            state = .stationary
            
            if state != previousState {
                persistState()
                emitEvent(previous: previousState, current: state)
            }
            switchToStationary()
        }
    }

    // MARK: - Location Feed

    /// Drive state machine transitions from a location fix speed.
    ///
    /// - Parameter speed: `CLLocation.speed` in m/s. Negative values (invalid)
    ///   are treated as 0 (stationary).
    public func onLocation(speed: Double) {
        guard isRunning else { return }

        let now = ProcessInfo.processInfo.systemUptime
        let effectiveSpeed = max(speed, 0)
        let previousState = state

        switch state {
        case .moving:
            handleMoving(speed: effectiveSpeed, now: now)
        case .slowing:
            handleSlowing(speed: effectiveSpeed, now: now)
        case .stationary:
            handleStationary(speed: effectiveSpeed, now: now)
        }

        lastFixTime = now

        // Persist and emit if state changed
        if state != previousState {
            persistState()
            emitEvent(previous: previousState, current: state)
        }
    }

    // MARK: - State Handlers

    private var slowingTimerWorkItem: DispatchWorkItem?

    private func handleMoving(speed: Double, now: TimeInterval) {
        if speed < speedMovingThreshold {
            let previousState = state
            state = .slowing
            lowSpeedCount = 1
            wakeCount = 0
            slowingStartTime = now
            NSLog("[SpeedMotion] MOVING -> SLOWING (speed=%.2f < threshold=%.2f)",
                  speed, speedMovingThreshold)
            startSlowingTimer()
            
            if state != previousState {
                persistState()
                emitEvent(previous: previousState, current: state)
            }
        }
    }

    private func startSlowingTimer() {
        stopSlowingTimer()
        let delay = TimeInterval(speedStationaryDelay)
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            let workItem = DispatchWorkItem { [weak self] in
                guard let self = self else { return }
                if self.state == .slowing {
                    NSLog("[SpeedMotion] SLOWING timer expired -> STATIONARY")
                    let previousState = self.state
                    self.state = .stationary
                    self.wakeCount = 0
                    self.lowSpeedCount = 0
                    self.switchToStationary()
                    
                    if self.state != previousState {
                        self.persistState()
                        self.emitEvent(previous: previousState, current: self.state)
                    }
                }
            }
            self.slowingTimerWorkItem = workItem
            DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: workItem)
        }
    }

    private func stopSlowingTimer() {
        DispatchQueue.main.async { [weak self] in
            self?.slowingTimerWorkItem?.cancel()
            self?.slowingTimerWorkItem = nil
        }
    }

    private func switchToStationary() {
        if stationaryTrackingMode == .geofences {
            delegate?.switchToStationaryGeofences()
        } else {
            delegate?.switchToStationaryPeriodic()
        }
    }

    private func handleSlowing(speed: Double, now: TimeInterval) {
        if speed >= speedMovingThreshold {
            // Back to moving
            let previousState = state
            state = .moving
            lowSpeedCount = 0
            stopSlowingTimer()
            NSLog("[SpeedMotion] SLOWING -> MOVING (speed=%.2f >= threshold=%.2f)",
                  speed, speedMovingThreshold)
                  
            if state != previousState {
                persistState()
                emitEvent(previous: previousState, current: state)
            }
            return
        }

        // Still slow — increment and check elapsed time
        lowSpeedCount += 1
        let elapsed = now - slowingStartTime
        if elapsed >= Double(speedStationaryDelay) {
            let previousState = state
            state = .stationary
            wakeCount = 0
            stopSlowingTimer()
            NSLog("[SpeedMotion] SLOWING -> STATIONARY (elapsed=%.0fs >= delay=%ds, lowCount=%d)",
                  elapsed, speedStationaryDelay, lowSpeedCount)

            switchToStationary()
            
            if state != previousState {
                persistState()
                emitEvent(previous: previousState, current: state)
            }
        }
    }

    private func handleStationary(speed: Double, now: TimeInterval) {
        if speed >= speedMovingThreshold {
            wakeCount += 1
            NSLog("[SpeedMotion] STATIONARY: wake fix (speed=%.2f, wakeCount=%d/%d)",
                  speed, wakeCount, speedWakeConfirmCount)
            if wakeCount >= speedWakeConfirmCount {
                state = .moving
                lowSpeedCount = 0
                wakeCount = 0
                NSLog("[SpeedMotion] STATIONARY -> MOVING (wakeConfirm reached)")
                delegate?.switchToContinuous()
            }
        } else {
            // Reset wake count on low-speed fix
            if wakeCount > 0 {
                NSLog("[SpeedMotion] STATIONARY: reset wakeCount (speed=%.2f < threshold)",
                      speed)
            }
            wakeCount = 0
        }
    }

    // MARK: - Persistence

    private func persistState() {
        stateManager.speedMotionState = state.rawValue
        stateManager.speedLowCount = lowSpeedCount
        stateManager.speedWakeCount = wakeCount
        stateManager.speedLastTransition = slowingStartTime
    }

    // MARK: - Event Emission

    private func emitEvent(previous: SpeedMotionState, current: SpeedMotionState) {
        let trackingMode: Int = (current == .stationary)
            ? (stationaryTrackingMode == .geofences ? 1 : 2) // geofences=1, periodic=2
            : 0 // continuous=0
        
        delegate?.emitSpeedMotionEvent(
            state: current.rawValue,
            previousState: previous.rawValue,
            trackingMode: trackingMode
        )
    }

    // MARK: - Config Loading

    /// Convenience to load config from a MotionConfig dictionary.
    public func loadConfig(from motionConfig: [String: Any]) {
        if let threshold = motionConfig["speedMovingThreshold"] as? Double {
            speedMovingThreshold = threshold
        }
        if let delay = motionConfig["speedStationaryDelay"] as? Int {
            speedStationaryDelay = delay
        }
        if let val = motionConfig["stationaryTrackingMode"] as? Int, let mode = StationaryTrackingMode(rawValue: val) {
            stationaryTrackingMode = mode
        }
        if let interval = motionConfig["stationaryPeriodicInterval"] as? Int {
            stationaryPeriodicInterval = interval
        }
        if let count = motionConfig["speedWakeConfirmCount"] as? Int {
            speedWakeConfirmCount = count
        }
    }
}
